plugins/janus_audiobridge.c

```diff
@@ -2329,6 +2354,120 @@ static json_t *janus_audiobridge_process_synchronous_request(janus_audiobridge_s
 		janus_mutex_unlock(&rooms_mutex);
 		JANUS_LOG(LOG_VERB, "Audiobridge room allowed list updated\n");
 		goto prepare_response;
+	} else if(!strcasecmp(request_text, "mute") || !strcasecmp(request_text, "unmute")) {
+		JANUS_LOG(LOG_VERB, "Attempt to mute a participant from an existing audiobridge room\n");
+		JANUS_VALIDATE_JSON_OBJECT(root, mute_parameters,
+			error_code, error_cause, TRUE,
+			JANUS_AUDIOBRIDGE_ERROR_MISSING_ELEMENT, JANUS_AUDIOBRIDGE_ERROR_INVALID_ELEMENT);
+		if(error_code != 0)
+			goto prepare_response;
+		json_t *room = json_object_get(root, "room");
+		json_t *id = json_object_get(root, "id");
+		gboolean muted = (!strcasecmp(request_text, "mute")) ? TRUE : FALSE;
+		guint64 room_id = json_integer_value(room);
+
+		janus_mutex_lock(&rooms_mutex);
+		janus_audiobridge_room *audiobridge = g_hash_table_lookup(rooms, &room_id);
+		if(audiobridge == NULL) {
+			janus_mutex_unlock(&rooms_mutex);
+			JANUS_LOG(LOG_ERR, "No such room (%"SCNu64")\n", room_id);
+			error_code = JANUS_AUDIOBRIDGE_ERROR_NO_SUCH_ROOM;
+			g_snprintf(error_cause, 512, "No such room (%"SCNu64")", room_id);
+			goto prepare_response;
+		}
+
+		janus_refcount_increase(&audiobridge->ref);
+		janus_mutex_lock(&audiobridge->mutex);
+		janus_mutex_unlock(&rooms_mutex);
+
+		/* A secret may be required for this action */
+		JANUS_CHECK_SECRET(audiobridge->room_secret, root, "secret", error_code, error_cause,
+			JANUS_AUDIOBRIDGE_ERROR_MISSING_ELEMENT, JANUS_AUDIOBRIDGE_ERROR_INVALID_ELEMENT, JANUS_AUDIOBRIDGE_ERROR_UNAUTHORIZED);
+		if(error_code != 0) {
+			janus_mutex_unlock(&audiobridge->mutex);
+			janus_refcount_decrease(&audiobridge->ref);
+			goto prepare_response;
+		}
+
+		guint64 user_id = json_integer_value(id);
+		janus_audiobridge_participant *participant = g_hash_table_lookup(audiobridge->participants, &user_id);
+		if(participant == NULL) {
+			janus_mutex_unlock(&audiobridge->mutex);
+			janus_refcount_decrease(&audiobridge->ref);
+			JANUS_LOG(LOG_ERR, "No such user %"SCNu64" in room %"SCNu64"\n", user_id, room_id);
+			error_code = JANUS_AUDIOBRIDGE_ERROR_NO_SUCH_USER;
+			g_snprintf(error_cause, 512, "No such user %"SCNu64" in room %"SCNu64, user_id, room_id);
+			goto prepare_response;
+		}
+		
+		participant->muted = muted;
```

Comment:
I'd add a further check before doing this. If we're trying to perform an operation that has already been done (i.e., trying to mute an already muted user, or trying to unmute a user that is not mute), then we should do nothing, and definitely not send another event to other participants; the only thing we should so is send a success back.

As such, I'd add something like

```
if(participant->muted == muted) {
	/* Nothing to do, just prepare response */
	response = json_object();
	json_object_set_new(response, "audiobridge", json_string("success"));

	/* Done */
	janus_mutex_unlock(&audiobridge->mutex);
	janus_refcount_decrease(&audiobridge->ref);
	goto prepare_response;
}
```

This makes the requests idempotent.

Reply:
Ah yes, sorry for that. Should've check it also.

Reply:
Hi @lminiero . I just added the condition.